src/ros_filter.cpp

```diff
@@ -2747,6 +2798,87 @@ namespace RobotLocalization
 
     return canTransform;
   }
+
+  template<typename T>
+  void RosFilter<T>::saveFilterState(FilterBase& filter)
+  {
+    FilterState state;
+    state.state_ = filter.getState();
+    state.estimate_error_covariance_ = filter.getEstimateErrorCovariance();
+    state.time_ = filter.getLastMeasurementTime();
+    filterStateHistory_.push_back(state);
+    ros::Time ts;
+    ts.fromSec(state.time_);
+    RF_DEBUG("\n---- RosFilter::saveFilterState for timestamp " << ts << " size = " << filterStateHistory_.size() << "----\n");
+  }
+
+  template<typename T>
+  bool RosFilter<T>::restoreFilterAndMeasurementsBefore(const double t)
+  {
+    RF_DEBUG("\n----- RosFilter::restoreFilterAndMeasurementsBefore (" << t << ") -----\n");
+    // Remove all filter states older than t. This yields either the latest
+    // filter state before t or an empty queue. In that case the smoothing
+    // window is too small and we cannot reset the filter.
+    //
+
+    int num_popped_filter_updates = 0;
+    while (!filterStateHistory_.empty()
+        && filterStateHistory_.back().time_ > t)
+    {
+      filterStateHistory_.pop_back();
+      num_popped_filter_updates++;
+    }
+    RF_DEBUG("\n---- Popped " << num_popped_filter_updates << " filter updates. Queue is empty/size? " << filterStateHistory_.empty() << "/" << filterStateHistory_.size() << "\n");
```

Comment:
Since empty is just size == 0, you can probably remove the empty printout. Also, there's an extra question mark after size.